### ISSUE TITLE
Princess Pathing NPE

### ISSUE THREAD

**Issue by BLOODWOLF333** [MegaMek/megamek#630]
Was testing out turrets on 0.43.3 and she was doing fine till this last turn when she just stopped moving the Stinger.  Looked in log and found an NPE I think involving her path decision.

Player name: BLOODWOLF
[PrincessPathingBug.zip](https://github.com/MegaMek/megamek/files/1278296/PrincessPathingBug.zip)

**Comment by BLOODWOLF333** [comment]
I also can deploy turrets on the ground (Long Tom turret in the saved game) but, I like it.

**Comment by kipstafoo** [comment]
This happened to me as well on Friday on the latest checked in code. My theory is it is related to princess ejecting disabled units. For my system, I do not have it create on-board entities to ejected pilots, it simply ejects and removes the unit.

**Comment by dericpage** [comment]
In BW's case, it's actually not the pathfinding.  Any time a turret is the target, the Compute class looks at the building it's emplaced in to determine its total CF + Armor.  If the turret is free-standing, then the building comes back as null.  The first question to be answered is whether or not free-standing turrets are legal (I have no idea).  If they are, then Compute needs to do a null-check.  If they are not, then the deployment rules for turrets needs to be changed.

**Comment by wildj79** [comment]
From Tactical Operations, p. 117

> *Gun Emplacements*
>
> Gun emplacements are single-hex buildings designed strictly to provide a weapons platform and protection for the crew manning those weapons. A variety of weapons and other special structure equipment can be mounted in a gun emplacement (see Building Construction, p. 128). Furthermore, unlike other building types, gun emplacements have minimal interior space, and so cannot be “entered” as a building by non-infantry units. Instead, they are largely treated as a stationary vehicular unit for purposes of stacking and making or receiving attacks, but as a building for tracking damage.

**Comment by arlith** [comment]
"Gun Emplacements" in MM are not TO Gun Emplacements.  As the quoted rules indicate, gun emplacements are actually a building type.  If you look at the code in MM, you can see that I think someone initially intended to implement Gun Emplacements, but what we have now is more of a kludge that doesn't really represent any TO rules.  The current implementation is more trying to implement the building construction rules, where you can construct a building that's basically a whole new unit.

This gets hairy for a lot of reasons, so basically the current MM implementation has `GunEmplacement` units that should only be deployed in buildings.  The `GunEmplacement` itself has no armor of internals, but it must exist within a building, and is destroyed if the building is destroyed.  This basically represents guns that can be built into buildings, like in the TO advanced building construction rules.

**Comment by arlith** [comment]
tldr; MM `GunEmplacement`s should be deployed inside of a building.  If they aren't, I'm assuming it's Princess deploying them, and that indicates there's a bug in the Princess deployment code.  This is kind of notorious, the server doesn't really check for legal deployments, instead the client does.  Sometimes the Princess client misses relevant checks and does bad things.

**Comment by pheonixstorm** [comment]
Princess is bad about stacking gun emplacements all in the same building at times iirc

**Comment by BLOODWOLF333** [comment]
Princess did not deploy the turrets I did, I NEVER allow her to deploy anything anymore.

**Comment by NickAragua** [comment]
I'll take a look at this tomorrow.

**Comment by NickAragua** [comment]
It appears that roads count as buildings for the purposes of player deployment checks only. I'll be removing that possibility.

**Comment by Dylan-M** [comment]
Fixed via #640
